app/src/main/java/com/ataulm/whatsnext/api/LetterboxdApiFactory.kt

```diff
@@ -0,0 +1,105 @@
+import java.util.*
+import java.util.concurrent.TimeUnit
+
+class LetterboxdApiFactory(
+        private val apiKey: String,
+        private val apiSecret: String,
+        private val tokensStore: TokensStore,
+        private val clock: Clock,
+        private val enableHttpLogging: Boolean
+) {
+
+    fun createRemote(): LetterboxdApi {
+        val okHttpClient = OkHttpClient.Builder()
+                .addNetworkInterceptor(AddApiKeyQueryParameterInterceptor(apiKey, clock))
+                .addNetworkInterceptor(AddAuthorizationHeaderInterceptor(apiSecret, tokensStore))
+
+        if (enableHttpLogging) {
+            val httpLoggingInterceptor = HttpLoggingInterceptor()
+            httpLoggingInterceptor.level = HttpLoggingInterceptor.Level.BODY
+            okHttpClient.addInterceptor(httpLoggingInterceptor)
+        }
+
+        return Retrofit.Builder()
+                .client(okHttpClient.build())
+                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
+                .addConverterFactory(GsonConverterFactory.create())
+                .baseUrl("https://api.letterboxd.com/api/v0/")
+                .build()
+                .create(LetterboxdApi::class.java)
+    }
+}
+
+private class AddApiKeyQueryParameterInterceptor(private val apiKey: String, private val clock: Clock) : Interceptor {
+
+    override fun intercept(chain: Interceptor.Chain): Response {
+        val url = chain.request().url.newBuilder()
+                .addQueryParameter("apikey", apiKey)
+                .addQueryParameter("nonce", UUID.randomUUID().toString())
+                .addQueryParameter("timestamp", TimeUnit.MILLISECONDS.toSeconds(clock.currentTimeMillis).toString())
+                .build()
+        val request = chain.request().newBuilder().url(url).build()
+        return chain.proceed(request)
+    }
+}
+
+private const val HEADER_AUTH = "Authorization"
+
+private class AddAuthorizationHeaderInterceptor(private val apiSecret: String, private val tokensStore: TokensStore) : Interceptor {
+
+    override fun intercept(chain: Interceptor.Chain): Response {
+        val builder = chain.request().newBuilder()
+        val originalUrl = chain.request().url
+
+        val signature = generateSignature(chain.request().method, originalUrl, chain.request().body)
+        if (originalUrl.requiresUserAuth()) {
+            val urlWithSignature = originalUrl.newBuilder().addQueryParameter("signature", signature).build()
+            builder.url(urlWithSignature)
+        } else {
+            builder.url(originalUrl)
+        }
+
+        if (originalUrl.requiresUserAuth()) {
+            builder.addHeader(HEADER_AUTH, "Bearer ${requireNotNull(tokensStore.token).accessToken}")
+        } else {
+            builder.addHeader(HEADER_AUTH, "Signature $signature")
+        }
+
+        return chain.proceed(builder.build())
+    }
+
+    private fun HttpUrl.requiresUserAuth(): Boolean {
+        return pathSegments.contains("me")
```

Comment:
It'd be cool to handle this at the API declarations, using annotations to mark the ones that require the access token, but it's not important.